lambdas/package/index.py

```diff
@@ -0,0 +1,186 @@
+"""
+Provide a virtual-file-system view of a package's logical keys.
+"""
+
+import io
+import json
+
+import boto3
+import pandas as pd
+
+from t4_lambda_shared.decorator import api, validate
+from t4_lambda_shared.utils import get_default_origins, make_json_response
+
+S3_DOMAIN_SUFFIX = '.amazonaws.com'
+
+SCHEMA = {
+    'type': 'object',
+    'properties': {
+        'bucket': {
+            'type': 'string'
+        },
+        'manifest': {
+            'type': 'string'
+        },
+        'access_key': {
+            'type': 'string'
+        },
+        'secret_key': {
+            'type': 'string'
+        },
+        'session_token': {
+            'type': 'string'
+        },
+        'prefix': {
+            'type': 'string'
+        },
+        'logical_key': {
+            'type': 'string'
+        }
+    },
+    'required': ['bucket', 'manifest', 'access_key', 'secret_key'],
+    'additionalProperties': False
+}
+
+
+class IncompleteResultException(Exception):
+    """
+    Exception indicating an incomplete response
+    (e.g., from S3 Select)
+    """
+
+
+def buffer_s3response(s3response):
+    """
+    Read a streaming response from s3 select into a
+    StringIO buffer
+    """
+    response = io.StringIO()
+    end_event_received = False
+    stats = None
+    for event in s3response['Payload']:
+        if 'Records' in event:
+            records = event['Records']['Payload'].decode()
+            response.write(records)
+        elif 'Progress' in event:
+            print(event['Progress']['Details'])
+        elif 'Stats' in event:
+            print(stats)
+        elif 'End' in event:
+            # End event indicates that the request finished successfully
+            end_event_received = True
+
+    if not end_event_received:
+        raise IncompleteResultException("Error: Received an incomplete response from S3 Select.")
+    response.seek(0)
+    return response
+
+
+def get_logical_key_folder_view(df):
+    """
+    Post process a set of logical keys to return only the
+    top-level folder view (a special case of the s3-select
+    lambda).
+    """
+    # matches all strings; everything before and including the first
+    # / is extracted
+    folder = pd.Series(df.logical_key.dropna().str.extract('([^/]+/?).*')[0].unique())
+    return dict(
+        prefixes=folder[folder.str.endswith('/')].sort_values().tolist(),
+        objects=folder[~folder.str.endswith('/')].sort_values().tolist()
+    )
+
+
+def get_s3_client(aws_access_key_id, aws_secret_access_key, aws_session_token):
+    """
+    Create an S3 Client using the provided credentials
+    """
+    session = boto3.Session(
+        aws_access_key_id=aws_access_key_id,
+        aws_secret_access_key=aws_secret_access_key,
+        aws_session_token=aws_session_token
+    )
+    s3_client = session.client('s3')
+    return s3_client
+
+
+def call_s3_select(s3_client, bucket, key, logical_key_prefix, detail=False):
```

Comment:
might as well start using type annotations in lambda: `bucket: string`, etc. and then `-> RETURN_TYPE`.